kernel/src/process.rs

```diff
@@ -340,96 +340,64 @@ impl<'a> Process<'a> {
-            mpu.set_mpu(0,
-                        text_start as u32,
-                        region_len,
-                        0,
-                        mpu::ExecutePermission::ExecutionPermitted,
-                        mpu::AccessPermission::ReadOnly);
+                        mpu::AccessPermission::ReadOnly) {
+            None =>
+                panic!("Infeasible MPU allocation. Base {:#x}, Length: {:#x}",
+                           text_start, text_len),
+            Some(region) => mpu.set_mpu(region),
         }
 
         let data_start = self.memory.as_ptr() as usize;
         let data_len = self.memory.len();
-        if data_len.count_ones() != 1 {
-            panic!("Tock MPU does not currently handle complex region sizes");
-        }
-        let data_region_len = math::PowerOfTwo::floor(data_len as u32);
-
 
-        // Data segment read/write/execute
-        mpu.set_mpu(1,
-                    data_start as u32,
-                    data_region_len,
-                    0,
-                    mpu::ExecutePermission::ExecutionPermitted,
-                    mpu::AccessPermission::ReadWrite);
+        match MPU::create_region(1, data_start, data_len,
+                        mpu::ExecutePermission::ExecutionPermitted,
+                        mpu::AccessPermission::ReadWrite) {
+            None =>
+                panic!("Infeasible MPU allocation. Base {:#x}, Length: {:#x}",
+                           data_start, data_len),
+            Some(region) => mpu.set_mpu(region)
+        }
 
         // Disallow access to grant region
-        let grant_size = unsafe {
+        let grant_len = unsafe {
             math::PowerOfTwo::ceiling(
                 self.memory.as_ptr().offset(self.memory.len() as isize) as u32 -
                     (self.kernel_memory_break as u32)
-            )
+            ).as_num::<u32>()
         };
         let grant_base = unsafe {
             self.memory
                 .as_ptr()
                 .offset(self.memory.len() as isize)
-                .offset(-(grant_size.as_num::<u32>() as isize))
+                .offset(-(grant_len as isize))
         };
-        mpu.set_mpu(2,
-                    grant_base as u32,
-                    grant_size,
-                    0,
-                    mpu::ExecutePermission::ExecutionNotPermitted,
-                    mpu::AccessPermission::PrivilegedOnly);
+
+        match MPU::create_region(2, grant_base as usize, grant_len as usize,
+                                 mpu::ExecutePermission::ExecutionNotPermitted,
+                                 mpu::AccessPermission::PrivilegedOnly) {
+            None =>
+                panic!("Infeasible MPU allocation. Base {:#x}, Length: {:#x}",
+                           grant_base as usize, grant_len),
+            Some(region) => mpu.set_mpu(region)
+        }
 
         // Setup IPC MPU regions
         for (i, region) in self.mpu_regions.iter().enumerate() {
```

Comment:
Actually, isn't this resolved by returning `None` from `create_region` if the region number is too high, and would be more general not to hard code the number of possible regions here as well?